### ISSUE TITLE
v4 prod bundle issue: UnhandledPromiseRejectionWarning: Error: Error replacing tree! Perhaps a syntax error was inserted (Current: ExpressionStatement -- New: SingleLineCommentTrivia).

### ISSUE THREAD

**Issue by radicand** [fuse-box/fuse-box#1645]
Perhaps related? #1634 

I am getting an error creating a production bundle in Fusebox v4 when using react-final-form. I've cloned the react-example repo and added the bare minimum to demonstrate the issue (a single form field). Repro repo here: https://github.com/radicand/react-example/tree/react-final-form-prod-repro

Dev bundle and HMR work fine.

Error printed:

```
$ npm run dist

> react-example@1.0.0 dist /Users/username/Projects/react-example
> ts-node -T fuse dist

  [ dist ] Starting

  ⚙  FuseBox 4.0.0-next.138
     Mode: production
     Entry: src/index.tsx

  source transform react-final-form/dist/react-final-form.es.js(node:66890) UnhandledPromiseRejectionWarning: Error: Error replacing tree! Perhaps a syntax error was inserted (Current: ExpressionStatement -- New: SingleLineCommentTrivia).
-- Details --
Path: /MyClass.tsx
Text: "...-hooks/exhaustive-deps\n  }, [decorators]); // warn about decorator changes\n  // istanbul ignore next\n // allow updatable config\n\n\n  useWhenValueChanges(debug, function () {\n    form.setConfig('debug',..."
Stack: Error: Error replacing tree! Perhaps a syntax error was inserted (Current: ExpressionStatement -- New: SingleLineCommentTrivia).
    at StraightReplacementNodeHandler.handleNode (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5659:19)
    at NodeHandlerHelper.handleForValues (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at DefaultParentHandler.handleNode (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5732:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5799:36)
    at NodeHandlerHelper.handleForValues (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5663:18)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5802:19)
    at NodeHandlerHelper.handleForValues (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at doManipulation (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:6460:15)
    at removeChildrenWithFormatting (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:6730:5)
    at removeStatementedNodeChildren (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:6806:5)
    at removeStatementedNodeChild (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:6803:5)
    at IfStatement.remove (/Users/username/Projects/react-example/node_modules/ts-morph/dist/ts-morph.js:13181:9)
    at setResult (/Users/username/Projects/react-example/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:63:18)
    at onIfStatement (/Users/username/Projects/react-example/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:77:17)
    at /Users/username/Projects/react-example/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:107:13
    at Array.forEach (<anonymous>)
    at Object.conditionUnwrapperProduction (/Users/username/Projects/react-example/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:105:17)
(node:66890) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:66890) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by nchanged** [comment]
thanks, yep, that's ts-morph issue. Maybe you could find a workaround for now? I am moving away from ts-morph

**Comment by radicand** [comment]
Thanks, I look forward to any resolution. Any tips for what I could attempt as a workaround?

**Comment by patrickmichalina** [comment]
@radicand it could also be TSC issue. I have seen 2 similar issues that are dependent on which version of typescript I am using. It only affects my Electron builds though.

**Comment by lamqson** [comment]
I am getting prod build after upgrading to MUI 4.4.3: 
```
source transform @material-ui/core/esm/Typography/index.js
  source transform @material-ui/core/esm/Typography/Typography.js
  source transform @material-ui/core/esm/InputLabel/index.js
  source transform @material-ui/core/esm/InputLabel/InputLabel.js
  source transform @material-ui/core/esm/FormControl/formControlState.js
  source transform @material-ui/core/esm/FormControl/useFormControl.js
  source transform @material-ui/core/esm/FormControl/FormControlContext.js
  source transform @material-ui/core/esm/FormLabel/index.js
  source transform @material-ui/core/esm/FormLabel/FormLabel.js
  source transform @material-ui/core/esm/MenuItem/index.js
  source transform @material-ui/core/esm/MenuItem/MenuItem.js
  source transform @material-ui/core/esm/ListItem/index.js
  source transform @material-ui/core/esm/ListItem/ListItem.js
(node:46629) UnhandledPromiseRejectionWarning: Error: Error replacing tree: Should not have children left over.
-- Details --
Path: /MyClass.tsx
Text: "... if (autoFocus) {\n      if (listItemRef.current) {\n        listItemRef.current.focus();\n      } else\n    }\n  }, [autoFocus]);\n  var children = React.Children.toArray(childrenProp);\n  var hasSecondaryA..."
Stack: Error: Error replacing tree: Should not have children left over.
    at DefaultParentHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5734:19)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5799:36)
    at NodeHandlerHelper.handleForValues (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5663:18)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5802:19)
    at NodeHandlerHelper.handleForValues (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5663:18)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5802:19)
    at doManipulation (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6460:15)
    at removeChildrenWithFormatting (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6730:5)
    at removeStatementedNodeChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6806:5)
    at removeStatementedNodeChild (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6803:5)
    at IfStatement.remove (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:13181:9)
    at setResult (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:63:18)
    at onIfStatement (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:77:17)
    at /Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:107:13
    at Array.forEach (<anonymous>)
    at Object.conditionUnwrapperProduction (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:105:17)
(node:46629) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:46629) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

✨  Done in 16.96s.
```
after downgrading to MUI 4.4.2 I start getting your error:
```
source transform react-final-form/dist/react-final-form.es.js
(node:48072) UnhandledPromiseRejectionWarning: Error: Error replacing tree! Perhaps a syntax error was inserted (Current: ExpressionStatement -- New: SingleLineCommentTrivia).
-- Details --
Path: /MyClass.tsx
Text: "...-hooks/exhaustive-deps\n  }, [decorators]); // warn about decorator changes\n  // istanbul ignore next\n // allow updatable config\n\n\n  useWhenValueChanges(debug, function () {\n    form.setConfig('debug',..."
Stack: Error: Error replacing tree! Perhaps a syntax error was inserted (Current: ExpressionStatement -- New: SingleLineCommentTrivia).
    at StraightReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5659:19)
    at NodeHandlerHelper.handleForValues (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at DefaultParentHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5732:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5799:36)
    at NodeHandlerHelper.handleForValues (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5663:18)
    at ParentFinderReplacementNodeHandler.handleNode (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5802:19)
    at NodeHandlerHelper.handleForValues (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5611:21)
    at ParentFinderReplacementNodeHandler.handleChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:5673:25)
    at doManipulation (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6460:15)
    at removeChildrenWithFormatting (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6730:5)
    at removeStatementedNodeChildren (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6806:5)
    at removeStatementedNodeChild (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:6803:5)
    at IfStatement.remove (/Users/sonlam/workspaces/octupi/raindrop/node_modules/ts-morph/dist/ts-morph.js:13181:9)
    at setResult (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:63:18)
    at onIfStatement (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:77:17)
    at /Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:107:13
    at Array.forEach (<anonymous>)
    at Object.conditionUnwrapperProduction (/Users/sonlam/workspaces/octupi/raindrop/node_modules/fuse-box/production/transformation/conditionUnwrapper.js:105:17)
(node:48072) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:48072) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by jcmoore0** [comment]
@nchanged  - Do you have any idea for workaround?  I changed to react-final-form this week and have the same issue with ts-morph.

**Comment by nchanged** [comment]
Could you guys pinpoint the exact line the one ts-morph doesnt like ? And hang on there guys, I am getting rid of that lib )

**Comment by nchanged** [comment]
Hey how about upgrading ts to the latest version ?

**Comment by radicand** [comment]
@nchanged I updated all the deps to latest versions in my test repo and it still fails on the same error. Here's the approximate line area where the error is printing:

```js
  }, [decorators]); // warn about decorator changes
  // istanbul ignore next

  if (process.env.NODE_ENV !== 'production') {
    // You're never supposed to use hooks inside a conditional, but in this
    // case we can be certain that you're not going to be changing your
    // NODE_ENV between renders, so this is safe.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useWhenValueChanges(decorators, function () {
      console.error('Form decorators should not change from one render to the next as new values will be ignored');
    }, shallowEqual);
  } // allow updatable config


  useWhenValueChanges(debug, function () {
    form.setConfig('debug', debug);
  });
```

I know nothing about how the parsing works, but is it possible that the comment at the end of the if block that is optimized out is causing it?

**Comment by jcmoore0** [comment]
Here is where it fails in MUI 4.4.3  (lastest version)

var useEnhancedEffect = typeof window === 'undefined' ? React.useEffect : React.useLayoutEffect;

@material-ui/core/esm/ListItem/ListItem.js 
```
useEnhancedEffect(function () {
    if (autoFocus) {
      if (listItemRef.current) {
        listItemRef.current.focus();
      } else if (process.env.NODE_ENV !== 'production') {
        console.error('Material-UI: unable to set focus to a ListItem whose component has not been rendered.');
      }
    }
  }, [autoFocus]);

**Comment by nchanged** [comment]
@radicand @jcmoore0 yes guys, that's totally ts-morph issue. I am unwrapping the conditions with
`process.env.NODE_ENV`  so you both have the same issue.  Can you try using pluginReplace on those files? Basically replace `process.env.NODE_ENV !== 'production'` `true` for dev `false` for production?

I am trying my best to deliver the best bundling experience and currently working on a custom typescript compiler that would solve many issues and make production builds very fast and memory efficient. But it requires a bit of time.

**Comment by jcmoore0** [comment]
@nchanged  - It still fails when replacing process.env.NODE_ENV !== 'production' false for production.   The hook is still parsed by ts-morph in below.  I would need to comment out the entire code block for it to work.
```
if (false) {
    // You're never supposed to use hooks inside a conditional, but in this
    // case we can be certain that you're not going to be changing your
    // NODE_ENV between renders, so this is safe.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useWhenValueChanges(decorators, function () {
      console.error('Form decorators should not change from one render to the next as new values will be ignored');
    }, shallowEqual);
  } // allow updatable config
```

**Comment by nchanged** [comment]
@jcmoore0 so is it because of the comment?

**Comment by jcmoore0** [comment]
@nchanged  - You are correct.  It is because of the comments inside the function.
```
// You're never supposed to use hooks inside a conditional, but in this
// case we can be certain that you're not going to be changing your
// NODE_ENV between renders, so this is safe.
// eslint-disable-next-line react-hooks/rules-of-hooks
```
The comment outside the function // allow ...  is ok.

**Comment by radicand** [comment]
@jcmoore0 in my testing, it's actually that comment outside the function on the same line as the closing parenthesis as @nchanged notes. I removed this and the bundle worked as expected in my environment. It's something of a hack to workaround this for now by manually rewriting the file in node_modules, but it can be done.

**Comment by nchanged** [comment]
@radicant you can use pluginReplace

I will submit an issue to ts-morph I think we can still solve it for the meantime

**Comment by jcmoore0** [comment]
@nchanged  - Thanks.  I added a repo  https://github.com/dsherret/ts-morph/issues/708

**Comment by dsherret** [comment]
Agree on not using ts-morph for a bundler. ts-morph is more for doing programmatic refactorings or code generation where things like comments are important. For a bundler you should be able to do AST manipulations so everything is quick then print out the final text from that changed AST. ts-morph currently reparses the source file between each manipulation (won't do that in the future, but it's a long term goal that would require a lot of cooperation with the TypeScript team to achieve).

**Issue Overview**

The issue here is that ts-morph is not properly handling removing `else if` blocks and that is a special case. You can see why in this screenshot (highlight shows the node location and what is being removed):

![image](https://user-images.githubusercontent.com/1609021/65842297-7e492380-e2f8-11e9-8283-92e4077d7169.png)

I just never came across this scenario.

The other error happens because it incorrectly changes the code from this:

```ts
// comment1
if (process.env.NODE_ENV !== 'production') {
} // comment2
test;
```

To this:

```ts
// comment1
// comment2
test;
```

For ts-morph, comment on a line by themselves are important because people may want to remove/insert around them when programmatically refactoring.

---

**Workaround**

So that said, as a workaround until this is fixed, do not use a `process.env.NODE_ENV !== 'production'` check in an else if block and do not add a comment on the last close brace.

**Comment by dsherret** [comment]
The issue in ts-morph is fixed in 4.1.1.

@nchanged note though that you do not want to be calling `#remove()` on an if statement because an if statement in the AST includes all the following else ifs. See the AST here: https://ts-ast-viewer.com/#code/JYMwBAFALgTgrgUwJRgN4FgBQWC+WD0+YUCAzlMAHYDmWCANqQmKJAPZQAWCMKG2mPJgZMW4CB2680WXEA

I've opened https://github.com/dsherret/ts-morph/issues/719 to develop a better api for dealing with if statements. In the meantime, you will probably want to change what you are doing to use an AST transform in some cases: https://ts-morph.com/manipulation/transforms

For example, to get rid of an if block and keep the child if statement (else statement), you could do the following:

```ts
ifStatement.transform(() => {
    return ifStatement.compilerNode.elseStatement!;
});
```

**Comment by nchanged** [comment]
thanks @dsherret 

@jcmoore0 @radicand is this an urgent fix you would like to see fixed? Or can you wait on that one?

**Comment by radicand** [comment]
@nchanged I have added a routine to hack out the comment in the react-final-form dist file in my node_modules so that it compiles for now. I can wait for an official fix later at this time.

**Comment by jcmoore0** [comment]
@nchanged  - I can wait for an official fix.  Thanks!

**Comment by charsleysa** [comment]
@radicand @jcmoore0 can you please confirm whether this issue occurs with the latest v4 alpha?

**Comment by radicand** [comment]
Testing against next-409, I no longer see this issue.
